### ISSUE TITLE
CUDA API interoperability

### ISSUE THREAD

**Issue by leofang** [numba/numba#4797]
Context: https://gitter.im/numba/numba?at=5da0d43a89acff6ff5db56eb

A question was raised recently for how to use a stream from CuPy in Numba's jit kernel. I also happened to run into the same question myself, and this was the workaround I had:
https://gist.github.com/leofang/4a043e5d94b4702d04fde2b9e7dcebbd
The `stream_cupy_to_numba()` function there allows the following simple usage:
```python
from numba import cuda
import cupy as cp

a = cp.random.random(100)
cp_stream = cp.cuda.stream.Stream()
nb_stream = stream_cupy_to_numba(cp_stream)
# ...define a @cuda.jit kernel...
kernel[grid, block, nb_stream](a)
```

However, it might be good for Numba to define a common interface to interoperate with other GPU libraries. First, we need a list of CUDA objects to be supported, and I think this list can be simple as this:
- device
- stream
- event

Second, in Numba one could initialize the CUDA object by accepting a 3rd party counterpart. Taking streams as example, this could be made valid I suppose:
```python
# passing the raw stream pointer to initialize a Numba stream  
# also need to tie the lifetime to the foreign object
nb_stream = cuda.stream(cp_stream.ptr, cp_stream)
```

(The title may not convey the essence of this task, so please feel free to reword it.)

**Comment by seibert** [comment]
This connects to a long standing idea / request I've had, which is that I would love to see an NVIDIA-managed simple Python wrapper around the CUDA Driver API.   (I've mentioned this to anyone who would listen.  Ping @kkraus14.)  Once we start talking about devices and streams (where I think different projects have no need for innovation), I think interfaces might be overkill, and dodge the problem that we have redundant implementations of these things everywhere.

A `cuda` Python package would create a standard set of classes to represent things like CUDA Devices, Streams, Events, etc and wrappers around the standard CUDA driver functions that operate on these things.  Then all the different CUDA Python projects could depend on this package, and have interoperable CUDA entities.  As a bonus, Numba could delete its wrappers around these things.

Note that this package ideally needs to be fairly small and basic: Smaller than Numba for sure, and even smaller than PyCUDA.  Wrappers based on ctypes would even allow the package to be pure Python, although Cython wrappers might be preferable for other reasons.

**Comment by leofang** [comment]
Thanks for sharing your thoughts, @seibert! I'm +1 for any possible official support like the `cuda` package you proposed. But I can imagine people would fight over whether it should be based on ctypes or cython. I like the latter more, but it might be more painful when it comes to packaging for distribution: looks like the code needs to be built against, and linked to, every alive CUDA version. (CuPy does this for CUDA 8.0 - 10.1...) Not sure how this can be handled gracefully so that downstream projects can quickly adapt it.

**Comment by seibert** [comment]
ctypes would have the advantage of being able to adapt at runtime to whatever CUDA is present on the system, so that would certainly be my preference.  This is how it works in Numba's wrapper, for example.  (And since the CUDA Driver API is a C API, the advantages of using Cython for wrapping C++ aren't relevant.)

**Comment by gmarkall** [comment]
In the short term, would be helpful if there were also a stream constructor that took a pointer (or int) to allow construction of a Numba `Stream` object for an existing stream without needing to use parts of Numba that are a bit "internal"?

**Comment by leofang** [comment]
> In the short term, would be helpful if there were also a stream constructor that took a pointer (or int) to allow construction of a Numba `Stream` object for an existing stream without needing to use parts of Numba that are a bit "internal"?

Yes, it'd be helpful, and is more or less what my `stream_cupy_to_numba()` did. Took me a while to figure out everything, so it'd be nice to have something similar like you said be built-in.

**Comment by gmarkall** [comment]
A relevant update: NVIDIA-supported wrappers for the CUDA Runtime and Driver APIs: https://developer.nvidia.com/cuda-python

**Comment by gmarkall** [comment]
Relabelling from "question" to "discussion" so that it doesn't go stale again - I plan to keep this open until Numba supports / uses the official CUDA Python bindings.
